### PR TITLE
Support multi-target

### PR DESCRIPTION
**Pull request by mateusduboli**
Adds support for [multi-target pattern](https://prometheus.io/docs/guides/multi-target-exporter/). Refers to #130 

Discussion points:
- Should we always enable the `scrapper` by default? Is this possibly a security problem?

### CODE REVIEW

**Comment by mateusduboli** [comment]
@grobie

**Comment by grobie** [comment]
Apologies for the radio silence @mateusduboli.

@SuperQ @roidelapluie So I see that the project's stance on multi-targets has changed? I found @SuperQ's comment from February suggesting exactly this change https://github.com/prometheus/memcached_exporter/issues/111#issuecomment-1026606633.

The change looks good to me in general. I believe users should explicitly opt-in into the new behavior though by setting a flag. It might be unexpected and even harmful for users upgrading to the newest version if the exporter can be suddenly used to make requests against arbitrary endpoints.

**Comment by SuperQ** [comment]
Yes, we've been adding multi-target exporter scrape support across the ecosystem to better support hosted/blackbox managed instances like Elasticache.

**Comment by SuperQ** [comment]
I don't think we've added any blockers / flags to any other exporter that disable multi-target by default. We typically recommend users that care implement auth/tls.

**Comment by mateusduboli** [comment]
Hi @grobie and @SuperQ, anything else that we can do to get this merged?

File: scraper/scraper.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+type Scraper struct {
+	logger  log.Logger
+	timeout time.Duration
+
+	scrapeCount  prometheus.Counter
+	scrapeErrors prometheus.Counter
+}
+
+func New(timeout time.Duration, logger log.Logger) *Scraper {
+	level.Debug(logger).Log("msg", "Started scrapper")
```
Comment (maxbrunet) [inline scraper/scraper.go:36]:
```suggestion
	level.Debug(logger).Log("msg", "Started scraper")
```

File: scraper/scraper.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			Name: "memcached_exporter_scrape_error_count",
+			Help: "Count of memcached exporter scape errors.",
+		}),
+	}
+}
+
+func (s *Scraper) Handler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		target := r.URL.Query().Get("target")
+		level.Debug(s.logger).Log("msg", "scrapping memcached", "target", target)
```
Comment (maxbrunet) [inline scraper/scraper.go:54]:
```suggestion
		level.Debug(s.logger).Log("msg", "scraping memcached", "target", target)
```

File: cmd/memcached_exporter/main_test.go [hunk 3]
```diff
@@ -75,32 +114,146 @@ OUTER:
...
+			t.Errorf("want metrics to include %q, have:\n%s", test, body)
+		}
+	}
+
+	cancel()
+
+	<-errc
+}
+
+func TestAcceptanceScrapper(t *testing.T) {
```
Comment (maxbrunet) [inline cmd/memcached_exporter/main_test.go:221]:
```suggestion
func TestAcceptanceScraper(t *testing.T) {
```

**Review by maxbrunet — COMMENTED** [review]
Found a recurring typo: `s/scrapp/scrap/` :slightly_smiling_face:

**scrape / scraping**: copy (data) from a website using a computer program.

**scrap / scrapping**: discard or remove from service (a retired, old, or inoperative vehicle, vessel, or machine), especially so as to convert it to scrap metal.

**Comment by mateusduboli** [comment]
@SuperQ @maxbrunet 
Any updates here?

File: scraper/scraper.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 The Prometheus Authors
```
Comment (SuperQ) [inline scraper/scraper.go:1]: Minor nit, when making new files, we set the current year. `2022`.

File: scraper/scraper_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 The Prometheus Authors
```
Comment (SuperQ) [inline scraper/scraper_test.go:1]: Same here.

**Review by SuperQ — CHANGES_REQUESTED** [review]
Minor nits.

Also, would you please update the README with documentation on how it's intended to work. I get it, but it would be nice to have docs right away.

File: scraper/scraper.go (hunk 4 above)
Comment (mateusduboli) [reply scraper/scraper.go]: Done

**Comment by mateusduboli** [comment]
@SuperQ done ✅

**Comment by pplu** [comment]
Can this get merged and released? It is a super-useful feature, and it looks like everything that was asked for has been done.

**Comment by mateusduboli** [comment]
@SuperQ @grobie any updates?

File: scraper/scraper.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	scrapeErrors prometheus.Counter
+}
+
+func New(timeout time.Duration, logger log.Logger) *Scraper {
+	level.Debug(logger).Log("msg", "Started scrapper")
+	return &Scraper{
+		logger:  logger,
+		timeout: timeout,
+		scrapeCount: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "memcached_exporter_scrape_count",
```
Comment (SuperQ) [inline scraper/scraper.go:41]:
```suggestion
			Name: "memcached_exporter_scrapes_total",
```

File: scraper/scraper.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+			Help: "Count of memcached exporter scapes.",
+		}),
+		scrapeErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "memcached_exporter_scrape_error_count",
```
Comment (SuperQ) [inline scraper/scraper.go:45]:
```suggestion
			Name: "memcached_exporter_scrape_errors_total",
```

**Review by SuperQ — CHANGES_REQUESTED** [review]
Please follow Prometheus best practices.

**Comment by chewrocca** [comment]
This feature seems helpful, but it has been open for several months without recent activity or approval. Is there a specific reason for the delay? Could someone help resolve any conflicts and get this merged?

**Comment by matthiasr** [comment]
We wrapped this up in #173.

**Comment by SuperQ** [comment]
Completed in https://github.com/prometheus/memcached_exporter/pull/173

**Pull request closed without merging**